graphix/sim/base_backend.py

```diff
@@ -1,10 +1,104 @@
 from __future__ import annotations
 
+import typing
+from dataclasses import dataclass
+
 import numpy as np
 
 import graphix.clifford
-import graphix.command
 import graphix.pauli
+import graphix.states
+from graphix.command import CommandKind
+from graphix.ops import Ops
+
+if typing.TYPE_CHECKING:
+    import collections
+
+    from graphix.pauli import Plane
+
+
+@dataclass
+class MeasurementDescription:
+    plane: Plane
+    angle: float
+
+
+class NodeIndex:
+    """
+    A class for managing the mapping between node numbers and qubit indices in the internal state
+    of the backend. This allows for efficient access and manipulation of qubit orderings
+    throughout the execution of a pattern.
+
+    Attributes:
+        __list (list): A private list of the current active node (labelled with integers).
+        __dict (dict): A private dictionary mapping current node labels (integers) to their corresponding qubit indices
+                       in the backend's internal quantum state.
+    """
+
+    def __init__(self) -> None:
+        self.__dict = dict()
+        self.__list = []
+
+    def copy(self) -> NodeIndex:
+        """
+        Creates a shallow copy of the NodeIndex instance, copying both the list and dictionary.
+        """
+        result = self.__new__(self.__class__)
+        result.__dict = self.__dict.copy()
+        result.__list = self.__list.copy()
+        return result
+
+    def __getitem__(self, index: int) -> int:
+        return self.__list[index]
+
+    def index(self, node: int) -> int:
+        """
+        Returns the qubit index associated with the specified node label.
+        """
+        return self.__dict[node]
+
+    def __iter__(self) -> collections.abc.Iterator[int]:
```

Comment:
Please add the corresponding `import`.
(I personally prefer single `Iterable` though)